crates/ruff_linter/src/rules/flake8_type_checking/rules/typing_only_runtime_import.rs

```diff
@@ -289,12 +291,24 @@ pub(crate) fn typing_only_runtime_import(
             continue;
         };
 
+        let mut binding_needs_future_import = true;
         if binding.context.is_runtime()
             && binding
                 .references()
                 .map(|reference_id| checker.semantic().reference(reference_id))
                 .all(|reference| {
-                    is_typing_reference(reference, &checker.settings().flake8_type_checking)
+                    let is_typing_ref =
+                        is_typing_reference(reference, &checker.settings().flake8_type_checking);
+
+                    // For a `from __future__ import annotations` to help, we should be in a type
+                    // definition and _not_ currently in a typing-only annotation. We also skip the
+                    // quoting related checks from `is_typing_reference`.
+                    binding_needs_future_import &= !is_typing_ref
+                        && reference.in_type_definition()
+                        && !reference.in_typing_only_annotation()
+                        && reference.in_runtime_evaluated_annotation();
```

Comment:
I think this `if` condition is worth splitting. It was already very long, but the fact that it now also mutates outer variable makes it rather surprising